engine/render/src/render/font_renderer.cpp

```diff
@@ -594,30 +607,21 @@ namespace dmRender
                 uint8_t is_compressed = *glyph_data++;
 
                 if (is_compressed) {
-
-                    uint32_t bytes_per_pixel;
-                    dmWebP::TextureEncodeFormat encode_format;
-                    switch (font_map->m_CacheFormat) {
-                        case dmGraphics::TEXTURE_FORMAT_RGB:        bytes_per_pixel = 3;
-                                                                    encode_format = dmWebP::TEXTURE_ENCODE_FORMAT_RGB888;
-                                                                    break;
-                        case dmGraphics::TEXTURE_FORMAT_RGBA:       bytes_per_pixel = 4;
-                                                                    encode_format = dmWebP::TEXTURE_ENCODE_FORMAT_RGBA8888;
-                                                                    break;
-                        case dmGraphics::TEXTURE_FORMAT_LUMINANCE:
-                        default:                                    bytes_per_pixel = 1;
-                                                                    encode_format = dmWebP::TEXTURE_ENCODE_FORMAT_L8;
-                    };
-
-                    dmWebP::Result result = dmWebP::DecodeCompressedTexture(glyph_data,
-                                                glyph_data_size,
-                                                font_map->m_CellTempData,
-                                                font_map->m_CacheCellWidth*font_map->m_CacheCellHeight*4, // the max size
-                                                tex_params.m_Width*bytes_per_pixel,
-                                                encode_format);
-
-                    if (result != dmWebP::RESULT_OK) {
-                        dmLogWarning("Failed to decompress glyph: %d", result);
+                    // When if came to choosing between the different algorithms, here are some speed/compression tests
+                    // Decoding 100 glyphs
+                    // lz4:     0.1060 ms  compression: 72%
+                    // deflate: 0.2190 ms  compression: 66%
+                    // png:     0.6930 ms  compression: 67%
+                    // webp:    1.5170 ms  compression: 55%
+
+                    FontGlyphInflaterContext deflate_context;
+                    deflate_context.m_Output = font_map->m_CellTempData;
+                    deflate_context.m_Cursor = 0;
+                    dmZlib::Result zlib_result = dmZlib::InflateBuffer(glyph_data, glyph_data_size, &deflate_context, FontGlyphInflater);
+                    if (zlib_result != dmZlib::RESULT_OK)
+                    {
+                        dmLogError("Failed to decompress glyph (%c)", g->m_Character);
+                        return;
```

Comment:
Since we removed the WebP support, we switched over to Deflate for font glyphs

Reply:
Will Deflate (or Inflate?..) decompress  textures faster than WebP? WebP was awfully slow on the HTML5 target, and especially on mobile browsers.

Reply:
We'll only use deflate/inflate for font glyphs. As you see in the comments in the code, it's ~7x faster than WebP.

Reply:
Transcoding the textures takes between 5-15ms (iirc), and the plan is to move that part from the loading into the graphics thread where we upload textures.

Reply:
Oh. I see it. Sorry, I was writing my answer from a phone yesterday and not seeing the comments :innocent:

Reply:
@JCash I'm not very familiar with Defold's threading model, but I'm worried it might affect async loading. Wouldn't doing work on the graphics thread mean that loading textures would block the rendering of a frame? AFAIK, in Vulkan, at least, you can load resources into VRAM on a separate thread without affecting the rendering thread.

I know we had issues with loading factories async causing skipped frames in the past (and I get that that may be unavoidable on OpenGL), but it should be as async as possible otherwise. Otherwise, we can't really stream textures mid-level and have to hide loading times behind static frames.

Reply:
Previously, we did WebP decompression on the loading thread. After that was completed, the texture was sent to be uploaded on a separate thread.

This "pre load" work is now done by the transcoder. And I see no reason to keep that work on the main thread?

Reply:
I don't think we can improve on the OpenGL context stalling from too much work.
What we can do is to at least not lock up the loader with the texture in question.

Also note that on desktop, we haven't had any hardware compression formats, so the uploads should not be significantly smaller (also making the opengl context to stall less).

Apart from that, I think we need to look into an actual use case to pinpoint any hitches in the loader, which might create skipped sound frames.

What compression settings did you have before? WebP?

Reply:
It was a long time ago while working on Interrogation. It happened on desktop (macOS and Windows) when loading big textures (4/8k). We wanted to split the loading of our cutscenes (big spine scenes) but we couldn't do it because loading the next spine scene would create a frame time spike.

I haven't tested that behaviour on the Vulkan backend (not sure how loading textures is implemented there). But yeah, I agree that we can't really do much on OpenGL, since contexts are not thread-safe (from what I remember), but for Vulkan it should be possible to upload on a different thread (I don't have Vulkan experience, but it certainly was possible with Metal when last I played with it). Is Defold already doing that?